core/src/main/java/org/apache/druid/math/expr/vector/VectorStringProcessors.java

```diff
@@ -32,32 +33,35 @@ public static <T> ExprVectorProcessor<T> concat(Expr.VectorInputBindingInspector
   {
     final ExprVectorProcessor processor;
     if (NullHandling.sqlCompatible()) {
-      processor = new StringOutStringsInFunctionVectorProcessor(
+      processor = new ObjectOutObjectsInFunctionVectorProcessor(
           left.buildVectorized(inspector),
           right.buildVectorized(inspector),
-          inspector.getMaxVectorSize()
+          inspector.getMaxVectorSize(),
+          ExpressionType.STRING
       )
       {
         @Nullable
         @Override
-        protected String processValue(@Nullable String leftVal, @Nullable String rightVal)
+        protected String processValue(@Nullable Object leftVal, @Nullable Object rightVal)
         {
           // in sql compatible mode, nulls are handled by super class and never make it here...
-          return leftVal + rightVal;
+          return leftVal + (String) rightVal;
```

Comment:
You could call `.toString()` and avoid ClassCastExceptions here?  Or, if it's important to do the cast, we should really do an instanceof check first so that we can generate a better error message than just `ClassCastException`